### PR TITLE
feat(RealitioWithAppeals): test file

### PR DESCRIPTION
**Pull request by unknownunknown1**

### CODE REVIEW

File: contracts/test/foreign-proxy-with-appeals.js [hunk 1]
```diff
@@ -0,0 +1,798 @@
...
+    .to.emit(homeProxy, "RequestCanceled")
+    .withArgs(questionID);
+
+  const newBalance = await requester.getBalance();
+  expect(newBalance).to.equal(oldBalance.add(5555), "Requester was not reimbursed correctly");
+
+  const arbitration = await foreignProxy.arbitrations(0);
+  expect(arbitration[0]).to.equal(0, "Status should be empty");
+  expect(arbitration[1]).to.equal(ZERO_ADDRESS, "Requester should be empty");
+  expect(arbitration[2]).to.equal(0, "Deposit should be empty");
```
Comment (hbarcelos) [inline contracts/test/foreign-proxy-with-appeals.js:193]: Split into 2 test cases grouped into a nested test suite:

```js
describe('When an arbitration is canceled', () => {
  it('should change the proxies state and emit the proper events', async () => {});
  it('reimburses the requester', async () => {});
});
```

File: contracts/test/foreign-proxy-with-appeals.js [hunk 2]
```diff
@@ -0,0 +1,798 @@
...
+    .to.emit(foreignProxy, "ArbitrationCanceled")
+    .withArgs(questionID);
+
+  const newBalance = await requester.getBalance();
+  expect(newBalance).to.equal(oldBalance.add(arbitrationCost), "Requester was not reimbursed correctly");
+
+  arbitration = await foreignProxy.arbitrations(0);
+  expect(arbitration[0]).to.equal(0, "Status should be empty");
+  expect(arbitration[1]).to.equal(ZERO_ADDRESS, "Requester should be empty");
+  expect(arbitration[2]).to.equal(0, "Deposit should be empty");
```
Comment (hbarcelos) [inline contracts/test/foreign-proxy-with-appeals.js:230]: Same as above.

File: contracts/test/foreign-proxy-with-appeals.js [hunk 3]
```diff
@@ -0,0 +1,798 @@
...
+
+  await expect(txFundAppeal)
+    .to.emit(foreignProxy, "Contribution")
+    .withArgs(arbitrationID, 0, 534, await crowdfunder1.getAddress(), 3500)
+    .to.emit(foreignProxy, "RulingFunded")
+    .withArgs(arbitrationID, 0, 534);
+
+  await expect(
+    foreignProxy.connect(crowdfunder1).fundAppeal(arbitrationID, 533, { value: appealCost })
+  ).to.be.revertedWith("Appeal fee is already paid.");
```
Comment (hbarcelos) [inline contracts/test/foreign-proxy-with-appeals.js:347]: Well, this is definitely a mouthful 😅.
It's really hard to follow what happens in this test case.

Things like:
```js
  txFundAppeal = foreignProxy
    .connect(crowdfunder1)
    .fundAppeal(arbitrationID, 533, { gasPrice: gasPrice, value: oneETH }); // Overpay to check that it's handled correctly.
```
definitely deserve their own dedicated test case with a proper description.

Also I don't fully get why would you split the contribution in 2 parts.

I see that you also take gas costs into account. By simplifying those test cases, you can leverage the [`changeBalance`](https://ethereum-waffle.readthedocs.io/en/latest/matchers.html#change-balance)  matcher from `ethereum-waffle`, which will automatically take care of gas costs, so you can focus on testing the logic instead of the plumbing.

File: contracts/test/foreign-proxy-with-appeals.js [hunk 4]
```diff
@@ -0,0 +1,798 @@
...
+  await foreignProxy.connect(requester).fundAppeal(arbitrationID, 41, { value: 8499 });
+
+  // Winner doesn't have to fund appeal in this case but let's check if it causes unexpected behaviour.
+  await foreignProxy.connect(crowdfunder2).fundAppeal(arbitrationID, 4, { value: oneETH });
+
+  await time.increase(appealTimeOut + 1);
+
+  await expect(foreignProxy.withdrawFeesAndRewards(arbitrationID, requesterAddress, 0, 50)).to.be.revertedWith(
+    "Dispute not resolved"
+  );
```
Comment (hbarcelos) [inline contracts/test/foreign-proxy-with-appeals.js:468]: Should be in its own test case "It should not allow to withdraw fees and rewards when the dispute is not resolved".

File: contracts/test/foreign-proxy-with-appeals.js [hunk 5]
```diff
@@ -0,0 +1,798 @@
...
+  newBalance = await crowdfunder1.getBalance();
+  expect(newBalance).to.equal(oldBalance.add(5294), "The balance of the crowdfunder1 is incorrect (4 ruling)"); // 7500 * 12000 / 17000.
+
+  oldBalance = await crowdfunder2.getBalance();
+  await foreignProxy.withdrawFeesAndRewards(arbitrationID, await crowdfunder2.getAddress(), 0, 4);
+  newBalance = await crowdfunder2.getBalance();
+  expect(newBalance).to.equal(oldBalance.add(705), "The balance of the crowdfunder2 is incorrect"); // 1000 * 12000 / 17000.
+});
+
+it("Should correctly withdraw appeal fees for multiple answers", async () => {
```
Comment (hbarcelos) [inline contracts/test/foreign-proxy-with-appeals.js:621]: This test case should be split into several ones.

File: contracts/test/foreign-proxy-with-appeals.js [hunk 6]
```diff
@@ -0,0 +1,798 @@
...
+    "The status should be Created."
+  );
+
+  await homeProxy.handleNotifiedRequest(questionID);
+  await expect(foreignProxy.connect(other).submitEvidence(arbitrationID, "text"))
+    .to.emit(foreignProxy, "Evidence")
+    .withArgs(arbitrator.address, arbitrationID, await other.getAddress(), "text");
+});
+
+it("Should make governance changes", async () => {
```
Comment (hbarcelos) [inline contracts/test/foreign-proxy-with-appeals.js:759]: Should also be split into multiple test cases, one for each governance method.

**Review by hbarcelos — CHANGES_REQUESTED** [review]
While tests seem to cover the majority of the contract, test cases are a bit difficult to follow because some of them test too much at once. Those should be split into more test cases.

**Comment by clesaege** [comment]
We do not do unit tests. What we want to test is interactions between multiples parts of the contract.
